src/io/export_stl.cc

```diff
@@ -39,121 +40,142 @@
 #endif
 
 namespace {
-
-std::string toString(const Vector3d& v)
+/* Define values for double-conversion library. */
+#define DC_BUFFER_SIZE (128)
+#define DC_FLAGS (double_conversion::DoubleToStringConverter::UNIQUE_ZERO)
+#define DC_INF NULL // Only finite values in STL outputs!
+#define DC_NAN NULL
+#define DC_EXP 'e'
+#define DC_DECIMAL_LOW_EXP (-6)
+#define DC_DECIMAL_HIGH_EXP (21)
+#define DC_MAX_LEADING_ZEROES (5)
+#define DC_MAX_TRAILING_ZEROES (0)
+
+std::string toString(const Vector3f& v)
 {
-  return STR(v[0], " ", v[1], " ", v[2]);
+  double_conversion::DoubleToStringConverter dc(
+    DC_FLAGS, DC_INF, DC_NAN, DC_EXP,
+    DC_DECIMAL_LOW_EXP, DC_DECIMAL_HIGH_EXP, DC_MAX_LEADING_ZEROES, DC_MAX_TRAILING_ZEROES
+  );
+
+  char buffer[DC_BUFFER_SIZE];
+
+  double_conversion::StringBuilder builder(buffer, DC_BUFFER_SIZE);
+  dc.ToShortestSingle(v[0], &builder);
+  builder.AddCharacter(' ');
+  dc.ToShortestSingle(v[1], &builder);
+  builder.AddCharacter(' ');
+  dc.ToShortestSingle(v[2], &builder);
+  builder.Finalize();
+
+  return buffer;
 }
 
-Vector3d toVector(const std::array<double, 3>& pt) {
-  return {pt[0], pt[1], pt[2]};
+int32_t flipEndianness(int32_t x) {
+  return 
+    ((x << 24) & 0xff000000) | ((x >> 24) & 0xff) |
+    ((x << 8) & 0xff0000) | ((x >> 8) & 0xff00);
 }
 
-Vector3d fromString(const std::string& vertexString)
-{
-  Vector3d v;
-  std::istringstream stream{vertexString};
-  stream >> v[0] >> v[1] >> v[2];
-  return v;
-}
+template <size_t N>
+void write_floats(std::ostream& output, const std::array<float, N>& data) {
+  static uint16_t test = 0x0001;
+  static bool isLittleEndian = *reinterpret_cast<char *>(&test) == 1;
 
-void write_vector(std::ostream& output, const Vector3f& v) {
-  for (int i = 0; i < 3; ++i) {
-    static_assert(sizeof(float) == 4, "Need 32 bit float");
-    float f = v[i];
-    char *fbeg = reinterpret_cast<char *>(&f);
-    char data[4];
+  if (isLittleEndian) {
+    output.write(reinterpret_cast<char *>(const_cast<float *>(&data[0])), N * sizeof(float));
+  } else {
+    std::array<float, N> copy(data);
 
-    uint16_t test = 0x0001;
-    if (*reinterpret_cast<char *>(&test) == 1) {
-      std::copy(fbeg, fbeg + 4, data);
-    } else {
-      std::reverse_copy(fbeg, fbeg + 4, data);
+    auto *ints = reinterpret_cast<int32_t *>(&copy[0]);
+    for (size_t i = 0; i < N; i++) {
+      ints[i] = flipEndianness(ints[i]);
     }
-    output.write(data, 4);
+
+    output.write(reinterpret_cast<char *>(&copy[0]), N * sizeof(float));
   }
 }
 
 
-uint64_t append_stl(const PolySet& ps, std::ostream& output, bool binary)
+uint64_t append_stl(const PolySet& polyset, std::ostream& output, bool binary)
 {
+  static_assert(sizeof(float) == 4, "Need 32 bit float");
+  // check if tessellation is needed
+  PolySet tmp(3);
+  bool needTessellation = !std::all_of(
```

Comment:
This is another refactoring opportunity, let the PolySet know if it's a triangle mesh or a polygon mesh. ..either through a flag or template args or smth.

Reply:
I also found the API a bit unwieldy. There is no point in passing the result as the second argument, we can just return a `std::shared_ptr<const PolySet>` or something. This way when the input is already a triangular mesh, we can just return the input.

Reply:
The second argument issue is resolved in my smart pointer PR